whylog/config/rule.py

```diff
@@ -4,13 +4,28 @@
 import six
 
 from whylog.config.parsers import RegexParserFactory
+from whylog.constraints.verifier import Verifier
 
 
 class Rule(object):
     def __init__(self, causes, effect, constraints):
         self._causes = causes
         self._effect = effect
         self._constraints = constraints
+        self._frequency_information = self._gather_causes_frequency_information()
+
+    def _gather_causes_frequency_information(self):
+        """
+        basing on self._causes and assumption that causes are sorted,
+        produces list of pairs: (parser name, number of occurrences of this parser)
+        """
+        causes_with_frequency_info = [(self._causes[0].name, 0)]
+        for cause in self._causes:
+            if cause.name == causes_with_frequency_info[-1][0]:
+                causes_with_frequency_info[-1] = (cause.name, causes_with_frequency_info[-1][1] + 1)
```

Comment:
this allocates a new tuple every time a parser matches. If you'd change it to a list, you could use `+= 1` on an indexed element to prevent reallocation. As far as I understand it, `itertools.groupby` combined with a len() on the output is what you should be doing here.